omegaconf/grammar_visitor.py

```diff
@@ -0,0 +1,348 @@
+import sys
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from antlr4 import TerminalNode
+
+from .errors import GrammarParseError
+
+if TYPE_CHECKING:
+    from .base import Node  # noqa F401
```

Comment:
I don't like that the visitor is aware of Node.
Seems like it's breaking the abstraction layers.

Can you explain why this visitors are returning Nodes sometimes?

Reply:
Something like this requires the interpolation to return the node:
```yaml
foo:
  x: 0
bar: ${foo}
```
Then trying to access `${bar}.x`.